### ISSUE TITLE
`dak.Array.__getitem__(slice)` is not implemented

### ISSUE THREAD

**Issue by jpivarski** [dask-contrib/dask-awkward#265]
For example,

```python
import dask_awkward as dak
import awkward as ak
a = dak.from_awkward(ak.Array(range(1000)), 10)
a[190:210].compute().show()
```

raises

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jpivarski/irishep/dask-awkward/src/dask_awkward/lib/core.py", line 959, in __getitem__
    return self._getitem_single(where)
  File "/home/jpivarski/irishep/dask-awkward/src/dask_awkward/lib/core.py", line 930, in _getitem_single
    raise DaskAwkwardNotImplemented(f"__getitem__ doesn't support where={where}.")
dask_awkward.utils.DaskAwkwardNotImplemented: __getitem__ doesn't support where=slice(90, 110, None).

If you would like this unsupported call to be supported by
dask-awkward please open an issue at:
https://github.com/dask-contrib/dask-awkward.
```

Slices are basic because this is how I'd test code on small samples: I'd get a Dask array, slice off the first 10 or so, and do all the work on that until everything seems to be working. Then, I'd remove the training wheels and let it run on the whole dataset.

For that to work, not only does `dak.Array.__getitem__(slice)` need to be implemented, but it needs to drop partitions that aren't included in the slice. As long as partition sizes are known, that's possible.

------------------------------

In `ak.Array.__getitem__` semantics, a simple slice can be in a tuple with other selectors, which complicates things. Since the first dimension is special, you can put `__getitem__` arguments into a canonical form by transforming

```python
array[before_basic, basic_slice, after_basic]
```

into

```python
array[basic_slice][before_basic, slice(None, None, None), after_basic]
```

where `before_basic` may be any field slices (`str` or iterable of `str`) and `after_basic` may be anything. Once you have it in this form, you only need to deal with the partitions in

```python
array[basic_slice]
```

and let the `ak.Array` deal with

```python
ak.Array.__getitem__((before_basic, Slice(None, None, None), after_basic))
```

in each partition individually.

---------------------------------------------

I tried some examples, and some of them are currently incorrect (as opposed to unimplemented):

```python
>>> a = dak.from_awkward(ak.Array({"x": range(1000)}), 10)
>>> a["x", 10:20].compute()
<Array [10, 11, 12, 13, 14, ..., 915, 916, 917, 918, 919] type='100 * int64'>
```

doesn't stop at 20, while

```python
>>> a[10:20, "x"].compute()
```

raises an error.

**Comment by jpivarski** [comment]
Tested in dask_awkward-2023.5.0 (same results as above).

**Comment by lgray** [comment]
I would mark this one as rather high priority. This is basic analysis functionality.

**Comment by martindurant** [comment]
Order of priorities?
- slice
- partition
- join (to single-partition)
- reduce axis=0

**Comment by lgray** [comment]
- reduce axis=0
- slice
- repartition
- join

**Comment by martindurant** [comment]
For slice, are we really looking for a head() function, start-stop slicing, or general start-stop-step (with negative values allowed) in the zeroth dimension?

**Comment by jpivarski** [comment]
In Awkward, start-stop slicing (negative indexes, no step) is implemented as a special case because it can always be purely a view.

That shouldn't be too difficult for Dask data with known partition sizes. But if the partition sizes are unknown, it would have to fetch them (or carry around a token that says, "I've sliced this," i.e. not eager).

With unknown partition sizes, a `head` function would require less fetching than general slicing, but it might suffer from an ambiguity in which asking for the first N elements returns N elements in the usual case, in which the first partition has more than N elements, but has surprises in corner cases.

**Comment by martindurant** [comment]
Correct, head(N) would be N rows from the first partition, if it has enough, or the whole of the first partition. This is the default behaviour is daskdataframe. darr[:N] would not give the same result for the second case.

**Comment by lgray** [comment]
From the user perspective, they'll want start-stop slicing in the zeroth dimension. Start-stop-skip on 0th I've not seen used much but would be a convenience for certain checks and statistical methods.

If with the skip is not too much to add it's probably worth it just to go for the general case.

**Comment by douglasdavis** [comment]
closed by #270 and #279
